File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Stack/stacksample.py
# Stack for temporary inventory actions (LIFO - e.g. undo last action)

class InventoryActionStack:
    def __init__(self):
        self.stack = []

    # Push: add a new inventory action to the top of the stack
    def push(self, action, brand, model, quantity):
        self.stack.append({
            "action": action,
            "brand": brand,
            "model": model,
            "quantity": quantity
        })

    # Pop: remove and return the last action (undo)
    def pop(self):
        if self.is_empty():
            return None
        return self.stack.pop()

    # Peek: view the top action without removing it
    def peek(self):
        if self.is_empty():
            return None
        return self.stack[-1]

    def is_empty(self):
        return len(self.stack) == 0

    def display(self):
        if self.is_empty():
            print("  (stack empty)")
            return
        for i, item in enumerate(reversed(self.stack)):
            print(f"  {i + 1}. [{item['action']}] {item['brand']} {item['model']}: {item['quantity']} units")


def main():
    # Stack for inventory actions (MotorPH)
    inventory_stack = InventoryActionStack()

    # Push new stock actions
    print("Pushing inventory actions onto stack...")
    inventory_stack.push("add", "Yamaha", "NMAX 155", 5)
    inventory_stack.push("add", "Honda", "Click 125", 10)
    inventory_stack.push("add", "Suzuki", "Burgman Street", 7)
    print("Stack (top to bottom):")
    inventory_stack.display()

    # Peek at top action
    print("\nPeek at top action:")
    top = inventory_stack.peek()
    print(f"  {top}")

    # Pop last action (undo)
    print("\nPop last action (undo)...")
    last_action = inventory_stack.pop()
    print(f"  Undid: {last_action}")
    print("Stack after pop:")
    inventory_stack.display()

    # Pop one more
    print("\nPop another action...")
    last_action = inventory_stack.pop()
    print(f"  Undid: {last_action}")
    print("Stack after pop:")
    inventory_stack.display()


if __name__ == "__main__":
    main()

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Stack/StackSample.java
package Stack;
import java.util.ArrayList;
import java.util.List;

/**
 * Stack Operations for Inventory Actions (LIFO - e.g. undo last action).
 */
public class StackSample {

    record ActionItem(String action, String brand, String model, int quantity) {}

    static class InventoryActionStack {
        private final List<ActionItem> stack = new ArrayList<>();

        void push(String action, String brand, String model, int quantity) {
            stack.add(new ActionItem(action, brand, model, quantity));
        }

        ActionItem pop() {
            if (stack.isEmpty()) {
                return null;
            }
            return stack.remove(stack.size() - 1);
        }

        ActionItem peek() {
            if (stack.isEmpty()) {
                return null;
            }
            return stack.get(stack.size() - 1);
        }

        boolean isEmpty() {
            return stack.isEmpty();
        }

        void display() {
            if (stack.isEmpty()) {
                System.out.println("  (stack empty)");
                return;
            }
            for (int i = stack.size() - 1; i >= 0; i--) {
                ActionItem item = stack.get(i);
                System.out.println("  " + (stack.size() - i) + ". [" + item.action() + "] "
                        + item.brand() + " " + item.model() + ": " + item.quantity() + " units");
            }
        }
    }

    public static void main(String[] args) {
        InventoryActionStack inventoryStack = new InventoryActionStack();

        System.out.println("Pushing inventory actions onto stack...");
        inventoryStack.push("add", "Yamaha", "NMAX 155", 5);
        inventoryStack.push("add", "Honda", "Click 125", 10);
        inventoryStack.push("add", "Suzuki", "Burgman Street", 7);
        System.out.println("Stack (top to bottom):");
        inventoryStack.display();

        System.out.println("\nPeek at top action:");
        System.out.println("  " + inventoryStack.peek());

        System.out.println("\nPop last action (undo)...");
        ActionItem lastAction = inventoryStack.pop();
        System.out.println("  Undid: " + lastAction);
        System.out.println("Stack after pop:");
        inventoryStack.display();

        System.out.println("\nPop another action...");
        lastAction = inventoryStack.pop();
        System.out.println("  Undid: " + lastAction);
        System.out.println("Stack after pop:");
        inventoryStack.display();
    }
}
